src/middlewared/middlewared/main.py

```diff
@@ -681,6 +687,23 @@ def _call_prepare(
             if methodobj._pass_app['message_id']:
                 args.append(message_id)
 
+        is_coroutine = asyncio.iscoroutinefunction(methodobj)
+        if (
+            serviceobj._config.namespace in ('zfs.resource')
+            or serviceobj._config.inject_lzh
+            or getattr(methodobj, '_inject_lzh', False)
+        ):
+            if is_coroutine:
+                raise RuntimeError("Thread local storage not valid for coroutines")
+
+            # inject the thread local storage object
+            # as the 1st positional argument to the
+            # method. NOTE: if the method is decorated
+            # with @api_method decorator, the `skip_args`
+            # MUST BE SET to `skip_args=1` so validation
```

Comment:
I don't understand, why `pass_app` doesn't require this kind of special attention, but `inject_lzh` does?

Also, naming might be a problem. We call them `pass_` and `inject_`, but they do essentially the same thing, right?

Reply:
Yeah, I misunderstood the argument index tracking logic that we wrote. The comment is invalid, I'll remove it. I'll go back to just doing `args.append` but I'll move this check to the beginning. Are you okay with that?

Reply:
> I'll go back to just doing args.append but I'll move this check to the beginning. Are you okay with that?

Yes, that'll be perfect. (I would prefer to have `lzh` argument after the `app` though in case both are passed, but that's my aesthetic preference).

Btw, at this time we have a lot of injected arguments (app, audit_callback, lzh, job). Should we add a check that checks the callee signature parameter names and ensures they are in the right order? If you don't want to do it here, I can do it in a separate PR after we merge this PR.